### PR TITLE
Add Tableau workbook refresh via webhooks

### PR DESCRIPTION
**Pull request by ramiz-bozai**
## What are you changing in this pull request and why?

A common workflow for data teams involves refreshing a Tableau workbook after rebuilding the underlying data.  dbt Cloud's new webhook functionality can be extended to this use case alongside the Tableau REST API.

## Checklist
- [x] Review the [Content style guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) and [About versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version) so my content adheres to these guidelines.
- [x] Add a checklist item for anything that needs to happen before this PR is merged, such as "needs technical review" or "change base branch."

Adding new pages (delete if not applicable):
- [x] Provide a unique filename for the new page

### CODE REVIEW

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
+---
+title: "Refresh a Tableau workbook with extracts when a job completes"
+id: webhooks-guide-zapier-refresh-tableau-workbook
+slug: zapier-refresh-tableau-workbook
+description: Use Zapier to trigger a Tableau workbook refresh
+---
+
+This guide will teach you how to refresh a Tableau workbook that leverages extracts when a dbt Cloud job has completed successfully and there is fresh data available. The integration will:
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]: sugg a link to tableau extracts
```suggestion
This guide will teach you how to refresh a Tableau workbook that leverages [extracts](https://help.tableau.com/current/pro/desktop/en-us/extracting_data.htm) when a dbt Cloud job has completed successfully and there is fresh data available. The integration will:
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+ - Receive a webhook notification in Zapier
+ - Trigger a refresh of a Tableau workbook
+
+## Prerequisites
+
+In order to set up the integration, you should have familiarity with:
+- [dbt Cloud Webhooks](/docs/deploy/webhooks)
+- Zapier
+- The [Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api.htm)
+- The [version](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_versions.htm#rest_api_versioning) of Tableau's REST API that is compatible with your server
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]:
```suggestion
To set up the integration, you need to be familiar with:

- [dbt Cloud Webhooks](/docs/deploy/webhooks)
- Zapier
- The [Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api.htm)
- The [version](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_versions.htm#rest_api_versioning) of Tableau's REST API that is compatible with your server 
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+In order to set up the integration, you should have familiarity with:
+- [dbt Cloud Webhooks](/docs/deploy/webhooks)
+- Zapier
+- The [Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api.htm)
+- The [version](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_versions.htm#rest_api_versioning) of Tableau's REST API that is compatible with your server 
+
+## Integration steps
+
+### 1. Obtain authentication credentials from Tableau
+Obtain a [Personal Access Token](https://help.tableau.com/current/server/en-us/security_personal_access_tokens.htm) from your Tableau Server/Cloud instance.  You'll need this to authenticate with the Tableau API.  In addition, you'll want to ensure that your Tableau workbook uses data sources that allow refresh access.  This setting is typically set when publishing.
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]:
```suggestion
To authenticate with the Tableau API, obtain a [Personal Access Token](https://help.tableau.com/current/server/en-us/security_personal_access_tokens.htm) from your Tableau Server/Cloud instance. In addition, make sure your Tableau workbook uses data sources that allow refresh access, which is usually set when publishing.
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+### 2. Create a new Zap in Zapier
+Use **Webhooks by Zapier** as the Trigger, and **Catch Raw Hook** as the Event. If you don't intend to [validate the authenticity of your webhook](docs/deploy/webhooks#validate-a-webhook) (not recommended!) then you can choose **Catch Hook** instead. 
+
+Press **Continue**, then copy the webhook URL.
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]: sugg clarifying intro to what user is doing here

```suggestion

To set up a webhook in Zapier, select **Webhooks by Zapier** as the Trigger and **Catch Raw Hook** as the Event. However, if you choose not to [validate the authenticity of your webhook](docs/deploy/webhooks#validate-a-webhook), which isn't recommended, you can choose **Catch Hook** instead. 

Press **Continue** and then copy the webhook URL. 
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+![Screenshot of the Zapier UI, showing the webhook URL ready to be copied](/img/guides/orchestration/webhooks/zapier-common/catch-raw-hook.png)
+
+### 3. Configure a new webhook in dbt Cloud
+See [Create a webhook subscription](/docs/deploy/webhooks#create-a-webhook-subscription) for full instructions. Your event should be **Run completed**, and you need to change the **Jobs** list to only contain any jobs whose completion should trigger a report refresh.
+
+Make note of the Webhook Secret Key for later.  Paste in the webhook URL obtained from Zapier in step 2 into the "Endpoint" field and test the endpoint.
+
+Once you've tested the endpoint in dbt Cloud, go back to Zapier and click **Test Trigger**, which will create a sample webhook body based on the test event dbt Cloud sent.
+
+The sample body's values are hard-coded and not reflective of your project, but they give Zapier a correctly-shaped object during development.
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]:
```suggestion
To set up a webhook subscription for dbt Cloud, follow the instructions in [Create a webhook subscription](/docs/deploy/webhooks#create-a-webhook-subscription). For the event, choose **Run completed** and modify the **Jobs** list to include only the jobs that should trigger a report refresh.

Remember to save the Webhook Secret Key for later. Paste in the webhook URL obtained from Zapier in step 2 into the **Endpoint** field and test the endpoint.

Once you've tested the endpoint in dbt Cloud, go back to Zapier and click **Test Trigger**, which will create a sample webhook body based on the test event dbt Cloud sent.

The sample body's values are hard-coded and not reflective of your project, but they give Zapier a correctly-shaped object during development. 
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+Zapier allows you to [store secrets](https://help.zapier.com/hc/en-us/articles/8496293271053-Save-and-retrieve-data-from-Zaps), which prevents your keys from being displayed in plaintext in the Zap code. You will be able to access them via the [StoreClient utility](https://help.zapier.com/hc/en-us/articles/8496293969549-Store-data-from-code-steps-with-StoreClient).
+
+This guide assumes the names for the secret keys are: `DBT_WEBHOOK_KEY`, `TABLEAU_SITE_URL`, `TABLEAU_SITE_NAME`, `TABLEAU_API_TOKEN_NAME`, and `TABLEAU_API_TOKEN_SECRET`. If you are using different names, make sure you update all references to them in the sample code.
+
+This guide uses a short-lived code action to store the secrets, but you can also use a tool like Postman to interact with the [REST API](https://store.zapier.com/) or create a separate Zap and call the [Set Value Action](https://help.zapier.com/hc/en-us/articles/8496293271053-Save-and-retrieve-data-from-Zaps#3-set-a-value-in-your-store-0-3).
+
+#### a. Create a Storage by Zapier connection
+If you haven't already got one, go to <https://zapier.com/app/connections/storage> and create a new connection. Remember the UUID secret you generate for later.
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]:
```suggestion
#### a. Create a Storage by Zapier connection 

Create a new connection at https://zapier.com/app/connections/storage if you don't already have one and remember the UUID secret you generate for later.
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+Choose **Run Python** as the Event. Run the following code: 
+```python 
+store = StoreClient('abc123') #replace with your UUID secret
+store.set('DBT_WEBHOOK_KEY', 'abc123') #replace with your dbt Cloud Webhook key
+store.set('TABLEAU_SITE_URL', 'abc123') #replace with your Tableau Site URL, inclusive of https:// and .com
+store.set('TABLEAU_SITE_NAME', 'abc123') #replace with your Tableau Site/Server Name
+store.set('TABLEAU_API_TOKEN_NAME', 'abc123') #replace with your Tableau API Token Name
+store.set('TABLEAU_API_TOKEN_SECRET', 'abc123') #replace with your Tableau API Secret
+```
+Test the step. You can delete this Action when the test succeeds. The key will remain stored as long as it is accessed at least once every three months.
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]:
```suggestion
#### b. Add a temporary code step

Choose **Run Python** as the Event and run the following code: 

```python 
store = StoreClient('abc123') #replace with your UUID secret
store.set('DBT_WEBHOOK_KEY', 'abc123') #replace with your dbt Cloud Webhook key
store.set('TABLEAU_SITE_URL', 'abc123') #replace with your Tableau Site URL, inclusive of https:// and .com
store.set('TABLEAU_SITE_NAME', 'abc123') #replace with your Tableau Site/Server Name
store.set('TABLEAU_API_TOKEN_NAME', 'abc123') #replace with your Tableau API Token Name
store.set('TABLEAU_API_TOKEN_SECRET', 'abc123') #replace with your Tableau API Secret
```
Test the step. You can delete this Action when the test succeeds. The key will remain stored as long as it is accessed at least once every three months.
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+### 4. Add a code action
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]: duplicate 4 so changing this to 5

```suggestion
### 5. Add a code action
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    "Accept": "application/json",
+    "Content-Type": "application/json",
+    "X-Tableau-Auth": auth_token
+}
+
+refresh_trigger = requests.post(refresh_url, data=json.dumps(refresh_data), headers=refresh_headers)
+return {"message": "Workbook refresh has been queued"}
+```
+
+### 5. Test and deploy
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]: Update step
```suggestion
### 6. Test and deploy
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+### 4. Add a code action
+Select **Code by Zapier** as the App, and **Run Python** as the Event. 
+
+In the **Set up action** area, add two items to **Input Data**: `raw_body` and `auth_header`. Map those to the `1. Raw Body` and `1. Headers Http Authorization` fields from the **Catch Raw Hook** step above.
+
+![Screenshot of the Zapier UI, showing the mappings of raw_body and auth_header](/img/guides/orchestration/webhooks/zapier-common/run-python.png)
+
+In the **Code** field, paste the following code, replacing `YOUR_STORAGE_SECRET_HERE` in the StoreClient constructor with the UUID secret you created when setting up the Storage by Zapier integration, and replacing the `workbook_name` and `api_version` variables to actual values.
+
+The code below will validate the authenticity of the request then obtain the workbook ID for the specified workbook name.  Next, the code will send a [`update workbook` command to the Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_ref_workbooks_and_views.htm#update_workbook_now) for the given workbook ID.
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]:
```suggestion
The following code validates the authenticity of the request and obtains the workbook ID for the specified workbook name.  Next, the code will send a [`update workbook` command to the Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_ref_workbooks_and_views.htm#update_workbook_now) for the given workbook ID.
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    "Content-Type": "application/json",
+    "X-Tableau-Auth": auth_token
+}
+
+refresh_trigger = requests.post(refresh_url, data=json.dumps(refresh_data), headers=refresh_headers)
+return {"message": "Workbook refresh has been queued"}
+```
+
+### 5. Test and deploy
+You can iterate on the Code step by modifying the code and then running the test again. When you're happy with it, you can publish your Zap.
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]:
```suggestion
To make changes to your code, you can modify it and test it again. When you're happy with it, you can publish your Zap.
```

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
+---
+title: "Refresh a Tableau workbook with extracts when a job completes"
```
Comment (mirnawong1) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]: non blocker but sugg shortening title
```suggestion
title: "Refresh Tableau workbook with extracts after a job finishes"
```

**Review by mirnawong1 — CHANGES_REQUESTED** [review]
hey @ramiz-bozai-dbt this looks awesome, thanks so much for opening this up! I've left some suggestions on my end and am happy to re-review it further!  let me know if you have any questions at all!

**Comment by ramiz-bozai** [comment]
> hey @ramiz-bozai-dbt this looks awesome, thanks so much for opening this up! I've left some suggestions on my end and am happy to re-review it further! let me know if you have any questions at all!

Thanks @mirnawong1 ! Added/revised changes as needed.  I'd appreciate one final review on this.

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+Choose **Run Python** as the Event and input the following code: 
+
+```python 
+store = StoreClient('abc123') #replace with your UUID secret
+store.set('DBT_WEBHOOK_KEY', 'abc123') #replace with your dbt Cloud Webhook key
+store.set('TABLEAU_SITE_URL', 'abc123') #replace with your Tableau Site URL, inclusive of https:// and .com
+store.set('TABLEAU_SITE_NAME', 'abc123') #replace with your Tableau Site/Server Name
+store.set('TABLEAU_API_TOKEN_NAME', 'abc123') #replace with your Tableau API Token Name
+store.set('TABLEAU_API_TOKEN_SECRET', 'abc123') #replace with your Tableau API Secret
+
```
Comment (joellabes) [inline website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]: Missing closing triple backticks here

**Review by joellabes — DISMISSED** [review]
lgtm!

File: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md (hunk 13 above)
Comment (ramiz-bozai) [reply website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md]: Thanks @joellabes .  Resolved.

**Review by mirnawong1 — APPROVED** [review]
this looks SOOOO good @ramiz-bozai-dbt ! lgtm as well and thank you for opening this up for users ✨

**Comment by mirnawong1** [comment]
will dismiss Joel's review since you've addressed this already and it's currently blocking your merge.

**Comment by mirnawong1** [comment]
will dismiss Joel's review since you've addressed this already and it's currently blocking your merge. good to merge for you @ramiz-bozai-dbt !!

**Pull request merged**